odoo/models.py

```diff
@@ -2756,74 +2756,96 @@ def _inherits_join_calc(self, alias, fname, query):
         :param query: query object on which the JOIN should be added
         :return: qualified name of field, to be used in SELECT clause
         """
+        sql = self._field_to_sql(alias, fname, query)
+        return self.env.cr.mogrify(sql).decode()
+
+    def _field_to_sql(self, alias: str, fname: str, query: (Query | None) = None) -> SQL:
+        """ Return an :class:`SQL` object that represents the value of the given
+        field from the given table alias, in the context of the given query.
+        The query object is necessary for inherited fields, many2one fields and
+        properties fields, where joins are added to the query.
+        """
+        full_fname = fname
         property_name = None
         if '.' in fname:
             fname, property_name = fname.split('.', 1)
             check_property_field_value_name(property_name)
```

Comment:
@std-odoo can you confirm that we could remove those checks?

Reply:
hum, it's touchy. At least can we keep a check in the write, so we keep clean names in database ? (for example to never allow someone to create properties like `qsd.qsd` which will behave weirdly when called in mapped for example)

As long as SQL.identifier always crash and never try to clean what we give to him (because currently it's possible to inject inside a query, but it crash later because of that so it's not usable), `IDENT_RE` should do the job indeed, but it's very touchy (so I still prefer to be paranoid and keep the check)

Reply:
Nevermind, I just found a sqli if we remove the checks `check_property_field_value_name` (I just return in the methods)



```py
def test_std(self):
    Model = self.env['test_new_api.message']
    self.discussion_1.attributes_definition = [{
        'name': "mypartner '",  # if we remove all checks on properties names
        'string': 'Partners',
        'type': 'many2one',
        'comodel': 'test_new_api.partner',
    }]
    self.messages.discussion = self.discussion_1
    self.env.flush_all()
    query = Model._search([])
    res = Model._inherits_join_calc("alias", "attributes.mypartner '", query)

Reply:
Actually, we should remove `_inherits_join_calc` instead. `mogrify` (in `_inherits_join_calc`) remove the security done by the SQL object in `_field_to_sql`.

Reply:
@std-odoo the value of `res` is the following text (when you `print` it)
```
COALESCE("alias"."attributes" -> 'mypartner \'' or #', 'false')
```
This looks like properly quoted SQL code to me.

Reply:
when I print it I get
```
COALESCE("alias"."attributes" -> 'mypartner \' or #', 'false')
```
![screenshot_spect](https://github.com/odoo/odoo/assets/54316226/c874ab67-a226-487a-b198-84c8c874843d)


maybe the `\` was confusing, you can do
```py
def test_std(self):
    Model = self.env['test_new_api.message']
    query = Model._search([])
    res = Model._inherits_join_calc("alias", "attributes.mypartner ' or #", query)
    print(res)
```

and it print
```
COALESCE("alias"."attributes" -> 'mypartner ' or #', 'false')
```

![image](https://github.com/odoo/odoo/assets/54316226/938037d0-0a6e-4911-8af4-3949a23f4529)